### DISCUSSION TITLE
Are dashboard and storefront really static applications?

### DISCUSSION THREAD

**Discussion by einar-hjortdal** [saleor/saleor#10471] [Q&A] [answered]
Hello, I am new to the world of frontend web development, please excuse me if I am confused.

On [this page](https://docs.saleor.io/docs/3.x/developer/running-saleor/docker) this is the information given `The dashboard and the storefront are static applications that do not require additional servers.` but in the [storefront repository's readme](https://github.com/saleor/react-storefront#motivation) this information is given: `modern & fast: Project is focusing on best practices for e-commerce, like: SSR & ISR, image optimization.`

My understand is that a static application does not require any server runtime to be served and working, but an application with SSR and ISR require some runtime to do the rendering. Am I mistaken?

Are dashboard and storefront really static applications?

**Comment by peelar** [comment, marked as answer]
Hello @Coachonko 👋 

First of all, thank you for your question. You describe yourself as someone who is "new to the world of frontend web development" but the question you ask is very insightful. I will do my best to answer it, but before you must know that these are not the details of Storefront/Dashboard themselves, but the technology we used for building them: [Next.js](https://nextjs.org/docs/getting-started). Blurred lines that separate different methods of building and rendering content have been confusing front-end developers for years now.

> My understand is that a static application does not require any server runtime to be served and working, but an application with SSR and ISR require some runtime to do the rendering. Am I mistaken?

Let's start with defining what a "static website" is. It is a website that's content (HTML and CSS) is generated during build-time and then those files end up being served to the user.

Most often, the data for your website lives somewhere else, so you have to fetch it. You would expect the process to be:
1. Fetch the data during build-time
2. Use the returned data to incorporate it inside the HTML, build it
3. Serve the built HTML and CSS to the user (no JavaScript!)

Unfortunately, it's not the way it works (at least not in Next.js). Even though the `getStaticProps` method is used, our website still needs JavaScript on the client side.

Next.js is not just a static site generator, it does whole sorts of different things, and some of them (like routing) run on the client side. In order for them to run correctly, the code for these methods needs to be available.

As far as I know, there are also some intricacies of how React's hydration works that complicate this matter a bit. No wonder new solutions emerge in the ecosystem, like [Astro](https://astro.build/). I can recommend [a cool video of Jack Herrington](https://youtu.be/PSzCtdM20Fc) describing what makes Astro websites truly static.

In the end, it's a trade-off between:
- building your website statically which makes it super fast (but you would have a full-page refresh when going from one page to another)
- making it rely on client-side JavaScript to improve the UX (in terms of routing, f.e.)

Next.js tries to reconcile both of those worlds 😄.

I hope that helps at least a bit. I can promise you that we indeed focus on the best practices that Next.js offers.

**Reply by einar-hjortdal** [reply]
Thank you for your answer, though I believe there is a misunderstanding of my question here.

As you may have guessed, I am mostly a backend developer and I do come from the days of templating languages and *traditional* websites. I do understand that a static web application requires a browser which can be considered a runtime. But the server that delivers the script to the browser should not need to have a runtime itself. 

This is my understanding of how static web applications work according to the explanation found [here](https://www.staticapps.org/articles/defining-static-web-apps/) and confirmed by your explanation 

> Unfortunately, it's not the way it works (at least not in Next.js). Even though the getStaticProps method is used, our website still needs JavaScript on the client side.

If for a static application to be static, there should not be server-side runtime, then why is the readme.md suggesting that the static applications dashboard and storefront have SSR & ISR?

> [Next.js](https://nextjs.org/docs/getting-started). Blurred lines that separate different methods of building and rendering content have been confusing front-end developers for years now.

I understand that next.js is a framework that allows to build web applications that use SSR and ISR, but are those also considered static applications? 

In practice, do the (default) dashboard and storefront require a javascript runtime on the server which delivers them to the browser, or can they be delivered to the browser by a web server that does not allow any server-side script execution (no SSR/ISR)?

**Reply by peelar** [reply]
AFAIK, it all comes down to the used definition of "static". Under the definition Next.js uses ([they say `getStaticProps` is a way to run Static Site Generation](https://nextjs.org/docs/basic-features/data-fetching/get-static-props)), Storefront is definitely static.

If "static" means to you that there is no server-side runtime, then perhaps you should be looking at [`next export`](https://nextjs.org/docs/advanced-features/static-html-export). It gives up on certain features of Next.js in favor of exporting static HTML files that don't need a server at all.

You are right to pay attention to this topic, though. We will try to improve our communication on that.

Is there any specific problem you are struggling with? Or is this simply out of curiosity?

**Reply by einar-hjortdal** [reply]
It is important to me as a system/network admin to understand how deployment can be done: if a javascript runtime on the server needs to be present (node.js/deno/bun/whatever), than deployment is more complex and potentially more costly than an actual static web application. Particularly in light of [this event](https://blog.heroku.com/next-chapter)

It seems to me that the definition of static web application is overly blurry on the next.js website. The definition that I have seen, and which seems widely accepted, is that the server lacks a javascript runtime. So if the storefront requires a server with a javascript runtime, then it is not a static web application.

Just to be clear, an example of a static web application is a "create react app" app without ReactDOMServer or similar libs.

The added complexity may not be appealing to users despite the added benefits of ssr. Of course, developing a static web app as frontend is an option, though before adopting Saleor one should understand what Saleor offers (and I didn't).
